File: paljs/src/lang/ast.ts
import type {Pos} from './parse';

export type ID = string;

export type IDMap<T> = Record<ID, T>;
export type IDSet = IDMap<boolean>;

export type Ann = {
  type: Expr;
  value?: Expr;
};

export type Ctx = IDMap<Ann>;

export const set: (_: ID[]) => IDSet = (ids) => Object.fromEntries(ids.map((id) => [id, true]));
export const empty = <T>(map: IDMap<T>) => Object.keys(map).length === 0;
export const union: <T>(..._: Record<ID, T>[]) => Record<ID, T> = Object.assign;
export const difference = (a: IDSet, b: IDSet): IDSet =>
  set(Object.keys(a).filter((k) => !(k in b)));
export const add = <T>(map: IDMap<T>, key: ID | undefined, value: NoInfer<T>): IDMap<T> =>
  key ? {...map, [key]: value} : map;
export const keys = <T>(map: IDMap<T>): IDSet => set(Object.keys(map));
export const without = <T>(map: IDMap<T>, key: ID | undefined) => {
  if (!key) return map;
  map = {...map};
  delete map[key];
  return map;
};

const freshen = (id: ID) =>
  id.replace(/[0-9]*$/, (tail) => (tail.length === 0 ? 0 : parseInt(tail) + 1).toString());
export const freshIn = (ids: IDSet, id: ID) => {
  while (id in ids) {
    id = freshen(id);
  }
  return id;
};

export const introduce = (ctx: Ctx, id: ID, type: Expr) => {
  const v = freshIn(keys(ctx), `_${id}0`);
  ctx = add(ctx, v, {type});
  return [ctx, newVar(v)] as const;
};

export const isRigid = (e: Expr): boolean => {
  if (isVar(e)) return !isHole(e);
  if (isApp(e)) return isRigid(e.fn) && isRigid(e.arg);
  return isRigid(e.argType) && isRigid(e.result);
};

export const isHole = (v: ID | Var | undefined) => {
  if (typeof v === 'object') {
    v = v.id;
  }
  return v?.startsWith('_') ?? true;
};

export type Program = Binding[][];

export type Binding = {
  id: ID;
  type?: Expr;
  value?: Expr;
};

export type Expr = {meta?: Pos} & (Var | App | Fn);

export type Var = {kind: 'var'; id: ID};
export type App = {
  kind: 'app';
  implicit: boolean;
  fn: Expr;
  arg: Expr;
};

export type FnKind = 'def' | 'type';
export type Fn = {
  kind: FnKind;
  implicit: boolean;
  argID?: ID;
  argType: Expr;
  result: Expr;
};

export const newVar = (id: ID): Var => ({kind: 'var', id});

export const isVar = (expr: Expr): expr is Var => {
  return expr.kind === 'var';
};

export const isApp = (expr: Expr): expr is App => {
  return expr.kind === 'app';
};

export const isFn = (expr?: Expr): expr is Fn => {
  return expr?.kind === 'def' || expr?.kind === 'type';
};

export const freeVars = (expr: Expr): IDSet => {
  if (isVar(expr)) return set([expr.id]);
  else if (isApp(expr)) return union(freeVars(expr.fn), freeVars(expr.arg));
  return union(
    freeVars(expr.argType),
    difference(freeVars(expr.result), expr.argID ? set([expr.argID]) : {})
  );
};

export const freeHoles = (expr: Expr): IDSet => {
  if (isVar(expr) && isHole(expr)) return set([expr.id]);
  else if (isVar(expr)) return set([]);
  else if (isApp(expr)) return union(freeHoles(expr.fn), freeHoles(expr.arg));
  return union(
    freeHoles(expr.argType),
    difference(freeHoles(expr.result), expr.argID ? set([expr.argID]) : {})
  );
};

export const occurs = (expr: Expr, v: Var): boolean => {
  if (isVar(expr)) return expr.id === v.id;
  else if (isApp(expr)) return occurs(expr.fn, v) || occurs(expr.arg, v);
  return occurs(expr.argType, v) || (expr.argID != v.id && occurs(expr.result, v));
};

export const substExpr = (expr: Expr, from: ID | undefined, to: Expr): Expr => {
  if (!from) return expr;
  if (isVar(expr)) {
    return expr.id === from ? to : expr;
  } else if (isApp(expr)) {
    return {
      ...expr,
      fn: substExpr(expr.fn, from, to),
      arg: substExpr(expr.arg, from, to),
    };
  } else {
    if (expr.argID == from) {
      return {...expr, argType: substExpr(expr.argType, from, to)};
    } else if (expr.argID == null) {
      return {
        ...expr,
        argType: substExpr(expr.argType, from, to),
        result: substExpr(expr.result, from, to),
      };
    }
    const newArgID = freshIn(difference(freeVars(to), set([expr.argID])), expr.argID);

    const result = substExpr(expr.result, expr.argID, newVar(newArgID));

    return {
      ...expr,
      argType: substExpr(expr.argType, from, to),
      result: substExpr(result, from, to),
    };
  }
};

export const alphaEquiv = (
  a: Expr,
  b: Expr,
  ctxA: (ID | undefined)[] = [],
  ctxB: (ID | undefined)[] = []
): boolean => {
  if (isVar(a) && isVar(b)) {
    return ctxA.indexOf(a.id) == ctxB.indexOf(b.id) && (ctxA.includes(a.id) || a.id == b.id);
  } else if (isApp(a) && isApp(b)) {
    return (
      a.implicit === b.implicit &&
      alphaEquiv(a.fn, b.fn, ctxA, ctxB) &&
      alphaEquiv(a.arg, b.arg, ctxA, ctxB)
    );
  } else if (isFn(a) && isFn(b)) {
    return (
      a.implicit === b.implicit &&
      a.kind === b.kind &&
      alphaEquiv(a.argType, b.argType, ctxA, ctxB) &&
      alphaEquiv(a.result, b.result, [a.argID, ...ctxA], [b.argID, ...ctxB])
    );
  } else {
    return false;
  }
};

const typeID = 'Type';
export const Type = newVar(typeID);

File: pal/lib/src/ast.dart
import 'package:pal/src/serialize.dart';

typedef ID = String;

extension Freshen on Iterable<ID> {
  static ID _freshen(ID id) => id.replaceFirstMapped(
        RegExp('[0-9]*\$'),
        (match) => ((int.tryParse(match[0]!) ?? 0) + 1).toString(),
      );

  ID freshen(ID id) {
    while (contains(id)) {
      id = _freshen(id);
    }
    return id;
  }
}

bool isHole(ID id) => id.startsWith('_');

extension type const IDMap<T>(Map<ID, T> map) {
  static IDMap<T> empty<T>() => const IDMap({});

  Iterable<ID> get keys => map.keys;
  bool get isEmpty => map.isEmpty;
  T? get(ID? key) => map[key];
  IDMap<T> add(ID? key, T value) => key != null ? IDMap({...map, key: value}) : this;
  IDMap<T> union(IDMap<T> other) => IDMap({...map, ...other.map});
  IDMap<T> without(ID? key) {
    final newMap = {...map};
    newMap.remove(key);
    return IDMap(newMap);
  }

  bool containsKey(ID key) => map.containsKey(key);

  bool equals(IDMap<T> other) {
    if (map.length != other.map.length) return false;
    for (final MapEntry(:key, :value) in map.entries) {
      if (other.get(key) != value) return false;
    }
    return true;
  }

  IDMap<T> restrict(Iterable<ID> ids) => IDMap({
        for (final id in ids)
          if (this.containsKey(id)) id: this.get(id) as T
      });

  IDMap<T> filter(bool Function(ID, T) f) => IDMap({
        for (final entry in map.entries)
          if (f(entry.key, entry.value)) entry.key: entry.value
      });
}

typedef Program = List<List<Binding>>;

class Binding {
  final ID id;
  final Expr? type;
  final Expr? value;

  const Binding(this.id, this.type, this.value) : assert(type != null || value != null);
}

sealed class Expr {
  final (int, int)? t;
  const Expr({this.t});

  @override
  String toString() => serializeExprIndent(80, withFullHoleNames: true);

  @override
  bool operator ==(Object other) => other is Expr && this.alphaEquiv(other);

  int _hashCode(List<String?> ctx) => switch (this) {
        Var v => Hash.all([!ctx.contains(v.id) ? v.id.hashCode : ctx.indexOf(v.id).hashCode]),
        App app => Hash.all([app.implicit.hashCode, app.fn._hashCode(ctx), app.arg._hashCode(ctx)]),
        Fn fn => Hash.all([
            fn.implicit.hashCode,
            fn.kind.hashCode,
            fn.argType._hashCode(ctx),
            fn.result._hashCode([fn.argID, ...ctx])
          ]),
      };

  @override
  int get hashCode => _hashCode(const []);
}

class Var extends Expr {
  final ID id;

  const Var(this.id, {super.t});
}

class App extends Expr {
  final bool implicit;
  final Expr fn;
  final Expr arg;

  const App(this.implicit, this.fn, this.arg, {super.t});
}

enum FnKind { Def, Typ }

class Fn extends Expr {
  static const Def = FnKind.Def;
  static const Typ = FnKind.Typ;

  final FnKind kind;
  final bool implicit;
  final ID? argID;
  final Expr argType;
  final Expr result;

  const Fn(this.implicit, this.kind, this.argID, this.argType, this.result, {super.t});
  const Fn.def(this.implicit, this.argID, this.argType, this.result, {super.t}) : kind = FnKind.Def;
  const Fn.typ(this.implicit, this.argID, this.argType, this.result, {super.t}) : kind = FnKind.Typ;
}

bool isRigid(Expr expr) {
  if (expr is Var) return !isHole(expr.id);
  if (expr is App) return isRigid(expr.fn) && isRigid(expr.arg);
  throw Error();
}

Set<ID> freeHoles(Expr expr) => expr.freeVars.where(isHole).toSet();

extension ExprOps on Expr {
  Set<ID> get freeVars => switch (this) {
        Var(:var id) => {id},
        Fn(:var argID, :var result, :var argType) =>
          result.freeVars.difference({argID}).union(argType.freeVars),
        App(:var fn, :var arg) => fn.freeVars.union(arg.freeVars),
      };

  bool occurs(Var v) => switch (this) {
        Var(:var id) => v.id == id,
        App(:var fn, :var arg) => fn.occurs(v) || arg.occurs(v),
        Fn(:var argID, :var argType, :var result) =>
          argType.occurs(v) || (v.id != argID && result.occurs(v))
      };

  Expr substExpr(ID? from, Expr to) {
    if (from == null) return this;
    switch (this) {
      case Var(:var id):
        return id == from ? to : this;
      case App(:var implicit, :var fn, :var arg):
        return App(implicit, fn.substExpr(from, to), arg.substExpr(from, to));
      case Fn(:var implicit, :var kind, :var argID, :var argType, :var result):
        if (argID == from) {
          return Fn(implicit, kind, argID, argType.substExpr(from, to), result);
        } else if (argID == null) {
          return Fn(implicit, kind, argID, argType.substExpr(from, to), result.substExpr(from, to));
        }
        var newArgID = to.freeVars.difference({argID}).freshen(argID);

        if (argID != newArgID) result = result.substExpr(argID, Var(newArgID));

        return Fn(
          implicit,
          kind,
          newArgID,
          argType.substExpr(from, to),
          result.substExpr(from, to),
        );
    }
  }

  bool alphaEquiv(Expr b, [List<String?> ctxA = const [], List<String?> ctxB = const []]) =>
      switch ((this, b)) {
        (Var(id: var a), Var(id: var b)) =>
          ctxA.indexOf(a) == ctxB.indexOf(b) && (ctxA.contains(a) || (a == b)),
        (App a, App b) => a.implicit == b.implicit &&
            a.fn.alphaEquiv(b.fn, ctxA, ctxB) &&
            a.arg.alphaEquiv(b.arg, ctxA, ctxB),
        (Fn a, Fn b) => a.implicit == b.implicit &&
            a.kind == b.kind &&
            a.argType.alphaEquiv(b.argType, ctxA, ctxB) &&
            a.result.alphaEquiv(b.result, [a.argID, ...ctxA], [b.argID, ...ctxB]),
        _ => false,
      };
}

const typeID = 'Type';
const Type = Var(typeID);

class Hash {
  static int combine(int hash, int value) {
    hash = 0x1fffffff & (hash + value);
    hash = 0x1fffffff & (hash + ((0x0007ffff & hash) << 10));
    return hash ^ (hash >> 6);
  }

  static int finish(int hash) {
    hash = 0x1fffffff & (hash + ((0x03ffffff & hash) << 3));
    hash = hash ^ (hash >> 11);
    return 0x1fffffff & (hash + ((0x00003fff & hash) << 15));
  }

  static int all(List<int> hashes) => finish(hashes.fold(0, combine));
}
